### PR TITLE
add basic Dexcom API metrics

### PR DESCRIPTION
**Pull request by ewollesen**
The oauth.Client.SendOauthRequest method is modified to accept a
[]request.ResponseInspect. One of these inspectors generates prometheus
metrics calls to track Dexcom API usage by number of calls, path, and status
code.

This is done in response to use seeing some 429 (too many requests) errors in our logs. With the visibility these metrics provide, we hope to better understand our usage of the Dexcom API, so that we can take steps to reduce our usage if necessary, or at least have a meaningful discussion with Dexcom about how much we'd like to use their API.

BACK-2820

### CODE REVIEW

File: dexcom/client/client.go
Comment (ewollesen) [inline dexcom/client/client.go:29]: This is a no-op, found via staticcheck.

File: client/client.go
Comment (toddkazakov) [inline client/client.go:83]: This change was the only red-flag for me in this PR, but I think it's ok, because `drainAndClose` is called in `handleResponse`. @darinkrauss is more familiar with the code than me, so I'll let him take a final a look.

Comment (ewollesen) [reply client/client.go:83]: Yeah, that change isn't strictly required, but as they're written before this, response inspectors that return errors, will cause the entire request to return an error, regardless of whether the request was otherwise successful. Any additional inspectors (like possibly the metrics inspector being added in this PR) might not run at all, causing inaccurate metrics.

Comment (darinkrauss) [reply client/client.go:83]: If we log the error (rather than failing) then the `drainAndClose` isn't necessary as it should be done in `handleResponse`.

Though logging (versus failing) changes the nature of the protocol, I don't see a problem with changing it since there isn't any previous usage.

Comment (darinkrauss) [reply client/client.go:83]: Oh, and the `continue` is unnecessary.

File: dexcom/client/client.go
Comment (darinkrauss) [reply dexcom/client/client.go:29]: When this was first implemented this was required because the Dexcom API did not correctly conform to the OAuth2 spec and this allowed the OAuth library to work with Dexcom. (I don't remember exactly what was broken, though.) Since we've updated this dependency since then and the function has become a no-op, and the Dexcom API appears to still work, it appears we can remove it. (Though, I suppose, in theory, this could be the cause of some of the Dexcom failures.)

File: request/inspector.go
Comment (darinkrauss) [inline request/inspector.go:12]: I don't think this is required. It is how we define the inspector and since we don't require this right now, I wouldn't make this claim. Just my two cents, though.

File: request/inspector.go
Comment (darinkrauss) [inline request/inspector.go:19]: Why? It seems like you are changing the intention of the interface without a need.

File: client/client.go
Comment (darinkrauss) [reply client/client.go:83]: Hmm, I'm not sure I agree with this change. If an inspector is important and shouldn't  be affected by errors from other inspectors, then it should be listed first.

However, I won't cause my disagreement to block this PR.

**Review by darinkrauss — DISMISSED** [review]
Approved even though I disagree with some of the changes.

File: request/inspector.go
Comment (ewollesen) [reply request/inspector.go:12]: This is just an attempt at documenting what users of this interface can expect.

Since there's no known ordering, any inspector that modifies the Response, could have an impact on later inspectors. As a result, warning folks to _not_ modify the Response seems like the safe path.

The comment aims to make it clear to someone writing their own ResponseInspector, that they probably shouldn't modify the Response, and if they do, they should be extra careful to do so in a way that doesn't interfere with other response inspectors.

File: request/inspector.go
Comment (darinkrauss) [inline request/inspector.go:34]: Is this `fmt.Fprintf` intentional?

**Review by darinkrauss — CHANGES_REQUESTED** [review]
I think there is an extraneous `fmt.Fprintf`?

File: request/inspector.go
Comment (ewollesen) [reply request/inspector.go:34]: Doh! Will fix.

**Review by darinkrauss — APPROVED** [review]
LGTM!

**Pull request merged**
